index.js

```diff
@@ -858,13 +858,143 @@ function normalizeQuery(inputQuery) {
       }
     }
   }
+
   // Deleted documents are kept around so that we can start their version from
   // the last version if they get recreated. Lack of a type indicates that a
   // snapshot is deleted, so don't return any documents with a null type
-  if (!query.$query._type) query.$query._type = {$ne: null};
+  if (deletedDocMightSatisfyQuery(query.$query)) {
+    query.$query._type = {$ne: null};
+  }
   return query;
 }
 
+// Might a deleted doc (one that contains {_type: null} and no other
+// fields) satisfy a query?
+//
+// Return true if it definitely can, or if we're not sure. (This
+// function is used as an optimization to see whether we can avoid
+// augmenting the query to ignore deleted documents)
+function deletedDocMightSatisfyQuery(query) {
+  // Any query with `{foo: value}` with non-null `value` will never
+  // match deleted documents (that are empty other than the `_type`
+  // field).
+  //
+  // This generalizes to additional classes of queries. Here’s a
+  // recursive description of queries that can't match a deleted doc:
+  // In general, a query with `{foo: X}` can't match a deleted doc
+  // if `X` is guaranteed to not match null or undefined. In addition
+  // to non-null values, the following clauses are guaranteed to not
+  // match null or undefined:
+  //
+  // * `{$in: [A, B, C]}}` where all of A, B, C are non-null.
+  // * `{$ne: null}`
+  // * `{$exists: true}`
+  // * `{$and: [A, B, C]}` where at least one of A, B, C is guaranteed to
+  //    not match null
+  // * `{$or: [A, B, C]}` where all of A, B, C are guaranteed to not match null
+  // * `{$gt: ...}`, `{gte: ...}`, `{$lt: ...}`, `{$lte: ...}`
+  //
+  // In addition, some queries that have `$and` or `$or` at the
+  // top-level can't match deleted docs:
+  // * `{$and: [A, B, C]}`, where at least one of A, B, C are queries
+  //   guaranteed to not match `{_type: null}`
+  // * `{$or: [A, B, C]}`, where all of A, B, C are queries guaranteed
+  //   to not match `{_type: null}`
+  //
+  // There are more queries that can't match deleted docs but they
+  // aren’t that common, e.g. ones using `$type` or bit-wise
+  // operators.
+  if ('$and' in query) {
+    for (var i = 0; i < query.$and.length; i++) {
+      if (!deletedDocMightSatisfyQuery(query.$and[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  if ('$or' in query) {
+    for (var i = 0; i < query.$or.length; i++) {
+      if (deletedDocMightSatisfyQuery(query.$or[i])) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  for (var prop in query) {
+    // When using top-level operators that we don't understand, play
+    // it safe
+    if (prop[0] === '$')
+      return true;
+
+    if (guaranteedToNotMatchNull(query[prop]))
+      return false;
+  }
+
+  return true;
+}
+
+function guaranteedToNotMatchNull(clause) {
```

Comment:
move to pattern where exhaustively search everything until return false and only return true prematurely